python/ray/data/_internal/plan.py

```diff
@@ -366,34 +370,33 @@ def schema(
             return self._schema
 
         schema = None
-
-        if self.has_computed_output():
-            schema = unify_block_metadata_schema(self._snapshot_bundle.metadata)
-        else:
-            schema = self._logical_plan.dag.aggregate_output_metadata().schema
-            if not schema and fetch_if_missing:
-                # For consistency with the previous implementation, we fetch the schema if
-                # the plan is read-only even if `fetch_if_missing` is False.
-
-                iter_ref_bundles, _, executor = self.execute_to_iterator()
-
+        if isinstance(self._logical_plan.dag, GuessMetadataMixin):
+            schema = self._logical_plan.dag.guess_schema()
+        if schema is None and fetch_if_missing:
+            # For consistency with the previous implementation, we fetch the schema if
+            # the plan is read-only even if `fetch_if_missing` is False.
+
+            iter_ref_bundles, _, executor = self.execute_to_iterator()
+            output_node: PhysicalOperator = executor._output_node[0]
+            with executor:
                 # Make sure executor is fully shutdown upon exiting
-                with executor:
-                    for ref_bundle in iter_ref_bundles:
-                        for metadata in ref_bundle.metadata:
-                            if metadata.schema is not None:
-                                schema = metadata.schema
-                                break
-
-        self._schema = schema
+                for _ in iter_ref_bundles:
```

Comment:
Schemas shouldn't be in the Operator hierarchy, but we can keep them inside OpState (for now).

Reply:
Let's consolidate this discussion in the other thread

https://github.com/ray-project/ray/pull/53454#discussion_r2127450273